### PR TITLE
Move stackdriver trace exporter to new interface and pdata

### PR DESCRIPTION
**Pull request by stevencl1013**
**Description:** Changes stackdriver trace exporter to use pdata and implement the new exporter interface rather than the old, and use the OT Go stackdriver exporter rather than the OC one. The metric exporter is just converting from pdata back to consumerdata and still using the OC exporter.

**Testing:** Modified the tests for stackdriver, data conversion, and factory to reflect changes.

### CODE REVIEW

File: exporter/stackdriverexporter/go.mod [hunk 1]
```diff
@@ -4,14 +4,16 @@ go 1.14
...
 require (
 	contrib.go.opencensus.io/exporter/stackdriver v0.13.1
+	github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/trace v0.2.1-0.20200710010320-8556f4b5bc41
 	github.com/census-instrumentation/opencensus-proto v0.2.1
-	github.com/golang/protobuf v1.3.5
+	github.com/golang/protobuf v1.4.2
 	github.com/shirou/gopsutil v2.20.4+incompatible // indirect
 	github.com/stretchr/testify v1.6.1
 	go.opencensus.io v0.22.3
 	go.opentelemetry.io/collector v0.5.1-0.20200721173458-f10fbf228f0e
```
Comment (stevencl1013) [inline exporter/stackdriverexporter/go.mod:13]: For some reason my test is failing with this version, but passes when I change it to `go.opentelemetry.io/collector v0.6.0`

Comment (tigrannajaryan) [reply exporter/stackdriverexporter/go.mod:13]: Likely related to https://github.com/open-telemetry/opentelemetry-collector/pull/1308

Comment (nilebox) [reply exporter/stackdriverexporter/go.mod:13]: See https://github.com/open-telemetry/opentelemetry-collector/pull/1308#discussion_r457895403
This is likely because of update to `github.com/golang/protobuf v1.4.2`
Try replacing `assert.Equal` with `proto.Equal` or `cmp.Diff` as in the PR linked above.

**Review by nilebox — CHANGES_REQUESTED** [review]

File: exporter/stackdriverexporter/go.mod (hunk 1 above)
Comment (stevencl1013) [reply exporter/stackdriverexporter/go.mod:13]: I think the test is failing because of the `go.opentelemetry.io/collector` version, not the `github.com/golang/protobuf` version. The test still fails even if I remove all uses of `assert.Equal`. Right now, it seems like `BatchWriteSpans` is never even being called. I'm seeing the error message
```
Failed to export to Stackdriver: rpc error: code = Canceled desc = context canceled
--- FAIL: TestStackdriverExport (10.00s)
    stackdriver_test.go:105: test timed out
```
Before I merged master into my branch, I was using
[this version](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/486/commits/b4f771c5b1cb7083766f8de56a55fbcaea8dbf3c#diff-37aff102a57d3d7b797f152915a6dc16R42) of the collector and I didn't experience this issue, and similarly when I change the version back to `0.5.0` or `0.6.0`, the test passes again.

Comment (nilebox) [reply exporter/stackdriverexporter/go.mod:13]: Please check again once #503 is merged.

**Comment by nilebox** [comment]
@stevencl1013 #503 is merged, please rebase and check if your tests succeed.

**Comment by stevencl1013** [comment]
Still fails. I think the problem has been occurring ever since [this PR](https://github.com/open-telemetry/opentelemetry-collector/pull/1386/files#diff-59023a9d8a3f84cc59727ba81aaa72a9L99) was merged. It seems that when it reaches [here](https://github.com/open-telemetry/opentelemetry-collector/pull/1386/files#diff-4e3d31918169c935e154a4f49dbbb47eR222-R231), the context is being canceled before the span is exported.

Before, the exporter's `ConsumeTraces` method would just call the `dataPusher` [here](https://github.com/open-telemetry/opentelemetry-collector/pull/1386/files#diff-59023a9d8a3f84cc59727ba81aaa72a9L99).

**Comment by nilebox** [comment]
@stevencl1013 I think I've found the root cause after debugging the test:
```
2020/07/27 16:11:20 Failed to export to Stackdriver: rpc error: code = Canceled desc = context canceled
    TestStackdriverExport: stackdriver_test.go:105: test timed out
```
Note the first error message. It comes from [cloudtrace.go#L221](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/7f25d8274c798f9537bf931bdd9b5c45de2d00a0/exporter/trace/cloudtrace.go#L221) (SDK exporter you have imported in this PR).

So we fail to export, because the context has been cancelled. The question is what is that context and why has it been cancelled early? The answer is related to #1386 as you suspected.

- `timeoutSender` creates a new context per request, and cancels it in [common.go#L229](https://github.com/open-telemetry/opentelemetry-collector/blob/db926fd7f488bfabe0a4558578b2d59c48cc56a8/exporter/exporterhelper/common.go#L229), right after export function is complete in [common.go#L231](https://github.com/open-telemetry/opentelemetry-collector/blob/db926fd7f488bfabe0a4558578b2d59c48cc56a8/exporter/exporterhelper/common.go#L231)
- SDK CloudTrace exporter is using bundler, which exports data **asynchronously**, as you've implemented it in https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/39
- This means that CloudTrace request is invoked in [trace.go#L143](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/7f25d8274c798f9537bf931bdd9b5c45de2d00a0/exporter/trace/trace.go#L143) **after** the context has been cancelled as I described above

Hopefully it's not too confusing?

The workaround is trivial: don't forward the context passed into collector exporter to SDK exporter, and create a new context instead.
In other words, in the `pushTraces` function do the following:
```go
        // NOTE: SDK Trace exporter exports data asynchronously via bundler, so we can't use the parent (synchronous) context, which gets cancelled immediately after this function is executed.
        ctx = context.Background() // create a new context to avoid cancellation
	for _, span := range spans {
		te.texporter.ExportSpan(ctx, span)
		goodSpans++
	}
```

Alternatively, we may make the use of bundler configurable, i.e. support "synchronous mode" which would've made more sense here.

/cc @james-bebbington

File: exporter/stackdriverexporter/stackdriver.go [hunk 2]
```diff
@@ -109,35 +136,53 @@ func newStackdriverExporter(cfg *Config) (*stackdriver.Exporter, error) {
...
+			return dropped, err
+		}
+	}
+	return dropped, nil
 }
 
-// pushTraceData is a wrapper method on StackdriverExporter.PushTraceSpans
-func (se *stackdriverExporter) pushTraceData(ctx context.Context, td consumerdata.TraceData) (int, error) {
+// pushTraces calls texporter.ExportSpan for each span in the given traces
+func (te *traceExporter) pushTraces(ctx context.Context, td pdata.Traces) (int, error) {
```
Comment (nilebox) [inline exporter/stackdriverexporter/stackdriver.go:166]:
```suggestion
func (te *traceExporter) pushTraces(_ context.Context, td pdata.Traces) (int, error) {
	// NOTE: SDK Trace exporter exports data asynchronously via bundler, 
 	// so we can't use the parent (synchronous) context, which gets cancelled 
	// immediately after this function is executed.
	ctx := context.Background() // create a new context to avoid cancellation
```

**Comment by james-bebbington** [comment]
> The workaround is trivial: don't forward the context passed into collector exporter to SDK exporter, and create a new context instead.

Great work finding this issue.

I think it would make more sense to put this fix into `operations-go` instead though, as the same issue could occur for other clients of that library with its current implementation. To avoid this kind of issue entirely, we should ensure we always create a new context at the time we start asynchronous work.

i.e. do not [pass context into bundler](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/master/exporter/trace/trace.go#L114); instead you can extract the spans from context first, and pass those in or do something similar.

**Comment by stevencl1013** [comment]
@nilebox Thanks for finding the cause of the issue! I committed the change and it seems to be working now, but I'll also look into modifying the `operations-go` repo as @james-bebbington mentioned.

File: exporter/stackdriverexporter/factory.go [hunk 3]
```diff
@@ -45,13 +46,19 @@ func (f *Factory) CreateDefaultConfig() configmodels.Exporter {
 }
 
 // CreateTraceExporter creates a trace exporter based on this config.
-func (f *Factory) CreateTraceExporter(logger *zap.Logger, cfg configmodels.Exporter) (component.TraceExporterOld, error) {
+func (f *Factory) CreateTraceExporter(
```
Comment (bogdandrutu) [inline exporter/stackdriverexporter/factory.go:49]: Use the new factory helpers

File: exporter/stackdriverexporter/go.mod [hunk 4]
```diff
@@ -4,11 +4,13 @@ go 1.14
 
 require (
 	contrib.go.opencensus.io/exporter/stackdriver v0.13.1
+	github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/trace v0.2.1-0.20200710010320-8556f4b5bc41
```
Comment (nilebox) [inline exporter/stackdriverexporter/go.mod:7]: The latest version is from Jul 29, 2020 (https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/commit/2752da7eaab799cbc680f91181a6af43866409b0), you need to update to it to include your changes on that side?

File: exporter/stackdriverexporter/go.mod [hunk 5]
```diff
@@ -4,11 +4,13 @@ go 1.14
...
 	github.com/census-instrumentation/opencensus-proto v0.3.0
 	github.com/golang/protobuf v1.4.2
 	github.com/stretchr/testify v1.6.1
 	go.opencensus.io v0.22.4
 	go.opentelemetry.io/collector v0.7.0
+  go.opentelemetry.io/otel v0.9.0
```
Comment (nilebox) [inline exporter/stackdriverexporter/go.mod:13]: Please fix indentation

File: exporter/stackdriverexporter/spandata.go [hunk 6]
```diff
@@ -17,293 +17,172 @@ package stackdriverexporter
...
-
-func protoEntriesToOCEntries(protoEntries []*tracepb.Span_Tracestate_Entry) []tracestate.Entry {
-	ocEntries := make([]tracestate.Entry, 0, len(protoEntries))
-	for _, protoEntry := range protoEntries {
-		var entry tracestate.Entry
-		if protoEntry != nil {
-			entry.Key = protoEntry.Key
-			entry.Value = protoEntry.Value
+	sc := apitrace.SpanContext{}
+	copy(sc.TraceID[:], span.TraceID())
```
Comment (nilebox) [inline exporter/stackdriverexporter/spandata.go:62]: Why do you have to copy slice items rather than just `sc.TraceID = span.TraceID()`?

Comment (nilebox) [reply exporter/stackdriverexporter/spandata.go:62]: I see that this is how it was done before, feel fee to keep it then.

File: exporter/stackdriverexporter/stackdriver.go [hunk 7]
```diff
@@ -21,57 +21,90 @@ import (
...
+func (*metricsExporter) Name() string {
+	return name
 }
 
-func (se *stackdriverExporter) Shutdown(context.Context) error {
-	se.exporter.Flush()
-	se.exporter.StopMetricsExporter()
+func (te *traceExporter) Shutdown(context.Context) error {
+	// uncomment when flush is implemented in trace exporter
+	// te.texporter.Flush()
```
Comment (nilebox) [inline exporter/stackdriverexporter/stackdriver.go:57]: https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/72 has been merged, you should be able to invoke flush now.

File: exporter/stackdriverexporter/stackdriver.go [hunk 8]
```diff
@@ -109,35 +136,57 @@ func newStackdriverExporter(cfg *Config) (*stackdriver.Exporter, error) {
...
 }
 
-// pushTraceData is a wrapper method on StackdriverExporter.PushTraceSpans
-func (se *stackdriverExporter) pushTraceData(ctx context.Context, td consumerdata.TraceData) (int, error) {
+// pushTraces calls texporter.ExportSpan for each span in the given traces
+func (te *traceExporter) pushTraces(_ context.Context, td pdata.Traces) (int, error) {
+	// NOTE: SDK Trace exporter exports data asynchronously via bundler,
+	// so we can't use the parent (synchronous) context, which gets cancelled
+	// immediately after this function is executed.
+	ctx := context.Background() // create a new context to avoid cancellation
```
Comment (nilebox) [inline exporter/stackdriverexporter/stackdriver.go:170]: This shouldn't be necessary anymore thanks to https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/74

File: go.mod [hunk 9]
```diff
@@ -41,7 +42,9 @@ require (
 	github.com/stretchr/testify v1.6.1
 	github.com/tcnksm/ghr v0.13.0
 	go.opentelemetry.io/collector v0.7.0
+  go.opentelemetry.io/otel v0.9.0 // indirect
```
Comment (nilebox) [inline go.mod:45]: Something is wrong with formatting

**Review by nilebox — COMMENTED** [review]
Looks good, just need a couple of final cleanups.

**Comment by nilebox** [comment]
@stevencl1013 the build is failing, please ping me when it's ready for another round of reviews.

**Review by nilebox — APPROVED** [review]
LGTM, thanks for patiently addressing issues and feedback :)

**Comment by nilebox** [comment]
@bogdandrutu Just noticed that my approval is not "green", could you add me to approvers in the contrib repo as well?

**Review by bogdandrutu — APPROVED** [review]
